### PR TITLE
LCH-4833: Run commands against included sites

### PR DESCRIPTION
**Pull request by ankitsingh0188**

### CODE REVIEW

File: src/Platform/AcquiaCloudPlatform.php [hunk 1]
```diff
@@ -182,14 +192,31 @@ public static function getEnvironmentQuestion(Config $config, AcquiaCloudClientF
...
   public function execute(Command $command, InputInterface $input, OutputInterface $output) : int {
+    if ($input->getOption('group') && $input->getOption('uri')) {
+      $helper = $command->getHelper('question');
+      $question = new ConfirmationQuestion('You have provided both the options, group as well as uri. We will ignore the uri option. Do you want to proceed (y/n)?', TRUE);
+      if (!$helper->ask($input, $output, $question)) {
+        return 1;
+      }
+    }
+
     $environments = new Environments($this->getAceClient());
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:203]: I would still start with this part. If there is no sites at all in the platform, then do not display the duplicated options message. Tell them first that something is wrong with the platform.

File: src/Platform/AcquiaCloudPlatform.php [hunk 2]
```diff
@@ -304,8 +331,65 @@ public function getActiveDomains(): array {
...
+   * @param OutputInterface $output
+   *   Output stream.
+   *
+   * @return array|int
+   *   Array containing filtered list of sites.
+   */
+  protected function filterSitesByGroup(string $group_name, array $sites, OutputInterface $output) {
+    $group_file = $this->groupingSitesFilePath();
+    try {
+      $file = Yaml::parseFile($group_file);
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:369]: It is just my opinion but `$file` is not so straightforward name for the value in this variable. Let's say `$group_config` or something similar.

File: src/Platform/AcquiaCloudPlatform.php [hunk 3]
```diff
@@ -304,8 +331,65 @@ public function getActiveDomains(): array {
...
+    if (!isset($file[$group_name])) {
+      $output->writeln('<error>Group name doesn\'t exists.</error>');
+      return 2;
+    }
+    elseif (empty($file[$group_name])) {
+      $output->writeln('<warning>No sites available in the groups. Exiting...</warning>');
+      return 3;
+    }
+
+    foreach ($sites as $key => $site) {
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:385]: Correct me if I am wrong, but would not be `array_intersect_key()` function could do this? It is possible I misunderstanding the goal here.

**Review by Ateee89 — CHANGES_REQUESTED** [review]

File: src/Platform/AcquiaCloudPlatform.php [hunk 4]
```diff
@@ -304,8 +331,65 @@ public function getActiveDomains(): array {
...
+   * @return string
+   *   Grouping file path.
+   */
+  protected function groupingSitesFilePath(): string {
+    $alias = $this->getAlias();
+    $dir_parts = static::PLATFORM_LOCATION;
+    array_unshift($dir_parts, getenv('HOME'));
+    $file = implode(DIRECTORY_SEPARATOR, $dir_parts) . "/{$alias}-groups.yml";
+
+    return $file;
```
[the remark this answers is not in the document]
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php]: Changed. Thanks.

File: src/Platform/AcquiaCloudPlatform.php [hunk 5]
```diff
@@ -304,8 +331,65 @@ public function getActiveDomains(): array {
...
+    catch (ParseException $exception) {
+      $output->writeln('<error>Unable to parse the YAML ' . $exception->getMessage() . '</error>', );
+      return 1;
+    }
+
+    if (!isset($file[$group_name])) {
+      $output->writeln('<error>Group name doesn\'t exists.</error>');
+      return 2;
+    }
+    elseif (empty($file[$group_name])) {
```
[the remark this answers is not in the document]
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php]: Changed. Thanks.

File: src/Platform/AcquiaCloudPlatform.php (hunk 1 above)
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php]: Changed. Thanks.

File: src/Platform/AcquiaCloudPlatform.php (hunk 2 above)
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php]: Changed. Thanks.

File: src/Platform/AcquiaCloudPlatform.php [hunk 6]
```diff
@@ -182,21 +192,38 @@ public static function getEnvironmentQuestion(Config $config, AcquiaCloudClientF
...
+        return 3;
+      }
+    }
+
     $sites = array_column($sites, 'uri');
     $args = $this->dispatchPlatformArgumentInjectionEvent($input, $sites, $command);
     $exit_code = 0;
     $vendor_paths = $this->get(self::ACE_VENDOR_PATHS);
+    $environments = new Environments($this->getAceClient());
     foreach ($this->get(self::ACE_ENVIRONMENT_DETAILS) as $application_id => $environment_id) {
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:224]: It won't work if we only filtering the values in `$sites` variable, cause the foreach running through the environment ids. We need to check that the given environment domain has in the sites or not and running the command or just call `continue` accordingly.

File: src/Platform/AcquiaCloudPlatform.php [hunk 7]
```diff
@@ -48,6 +51,13 @@ class AcquiaCloudPlatform extends PlatformBase implements PlatformSitesInterface
 
   public const ACE_SITE_HTTP_PROTOCOL = 'acquia.cloud.environment.sites';
 
+  public const ACE_INCLUDED_SITES = 'acquia.cloud.environment.include';
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:54]: Do we need this constant?

File: src/Platform/AcquiaCloudPlatform.php [hunk 8]
```diff
@@ -182,21 +192,42 @@ public static function getEnvironmentQuestion(Config $config, AcquiaCloudClientF
...
+    }
+
+    if ($group_name) {
+      $sites = $this->filterSitesByGroup($group_name, $sites, $output);
+      if (empty($sites)) {
+        return 3;
+      }
+    }
+
+    $sites_uri = array_column($sites, 'uri');
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:219]: Why do we need this here?

File: src/Platform/AcquiaCloudPlatform.php [hunk 9]
```diff
@@ -182,21 +192,42 @@ public static function getEnvironmentQuestion(Config $config, AcquiaCloudClientF
...
+    $args = $this->dispatchPlatformArgumentInjectionEvent($input, $sites_uri, $command);
     $exit_code = 0;
     $vendor_paths = $this->get(self::ACE_VENDOR_PATHS);
+    $environments = new Environments($this->getAceClient());
     foreach ($this->get(self::ACE_ENVIRONMENT_DETAILS) as $application_id => $environment_id) {
+      if (!in_array($environment_id, array_keys($sites))) {
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudPlatform.php:225]:
```suggestion
      if (!in_array($environment_id, array_keys($sites), TRUE)) {
```

File: src/Platform/AcquiaCloudMultiSitePlatform.php [hunk 10]
```diff
@@ -50,25 +50,42 @@ public function execute(Command $command, InputInterface $input, OutputInterface
...
+      }
+    }
+
+    if ($group_name) {
+      $sites = $this->filterSitesByGroup($group_name, $sites, $output);
+      if (empty($sites)) {
+        return 3;
+      }
+    }
+    elseif ($input_uri) {
```
Comment (Ateee89) [inline src/Platform/AcquiaCloudMultiSitePlatform.php:77]: Do not use `elseif` unless it is unavoidable!! I think we can get rid off it here.

**Review by Ateee89 — CHANGES_REQUESTED** [review]

File: src/Platform/AcquiaCloudPlatform.php [hunk 11]
```diff
@@ -182,21 +185,44 @@ public static function getEnvironmentQuestion(Config $config, AcquiaCloudClientF
...
     }
 
+    $group_name = $input->getOption('group');
     $input_uri = $input->getOption('uri');
-    $sites = array_column($sites, 'uri');
-    $args = $this->dispatchPlatformArgumentInjectionEvent($input, $sites, $command);
+
+    if ($group_name && $input_uri) {
+      $helper = $command->getHelper('question');
+      $question = new ConfirmationQuestion('You have provided both the options, group as well as uri. We will ignore the uri option. Do you want to proceed (y/n)?', TRUE);
```
Comment (amangrover90) [inline src/Platform/AcquiaCloudPlatform.php:199]: This is problematic when a command runs internally through PlatformCommand Executioner in which case user isn't shown this question and execution kind of halts without informing anything to the user.

File: src/Platform/AcquiaCloudMultiSitePlatform.php [hunk 12]
```diff
@@ -50,25 +50,45 @@ public function execute(Command $command, InputInterface $input, OutputInterface
...
       return 1;
     }
 
+    $group_name = $input->getOption('group');
     $input_uri = $input->getOption('uri');
-    if ($input_uri) {
+
+    if ($group_name && $input_uri) {
+      $helper = $command->getHelper('question');
+      $question = new ConfirmationQuestion('You have provided both the options, group as well as uri. We will ignore the uri option. Do you want to proceed (y/n)?', TRUE);
```
Comment (amangrover90) [inline src/Platform/AcquiaCloudMultiSitePlatform.php:65]: Same for this one.

File: src/Platform/AcquiaCloudPlatform.php (hunk 11 above)
Comment (amangrover90) [reply src/Platform/AcquiaCloudPlatform.php:199]: Let's ignore group and only use uri in this case without any choice question.

**Review by amangrover90 — CHANGES_REQUESTED** [review]
Left some comments which impact command execution.

File: src/Platform/AcquiaCloudPlatform.php (hunk 11 above)
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php:199]: Let's ignore group and only use uri in this case without any choice question.

Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudPlatform.php]: Done.

File: src/Platform/AcquiaCloudMultiSitePlatform.php (hunk 12 above)
Comment (ankitsingh0188) [reply src/Platform/AcquiaCloudMultiSitePlatform.php]: Done.

**Pull request closed without merging**